LIS3MDL.device.nut

```diff
@@ -0,0 +1,236 @@
+
+class LIS3MDL {
+
+    static VERSION = [1,0,0];
+
+    // External constants
+    static AXIS_X = 0x80;
+    static AXIS_Y = 0x40;
+    static AXIS_Z = 0x20;
+    // Note that if INTERRUPT_ISACTIVEHIGH is used to generate a wake signal, significant power may be wasted.  See README for details.
+    static INTERRUPT_ISACTIVEHIGH = 0x04;
+    static INTERRUPT_DONTLATCH = 0x02;
+    static DATA_RATE_FAST = -1;
+    static CONVERSION_TYPE_CONTINUOUS = 0x00;
+    static CONVERSION_TYPE_SINGLE = 0x01;
+
+    // Internal constants
+    static REG_ADDR_OUT_X_L = 0x28;
+    static REG_ADDR_OUT_Y_L = 0x2A;
+    static REG_ADDR_OUT_Z_L = 0x2C;
+    static REG_CTL_1 = 0x20;
+    static REG_CTL_2 = 0x21;
+    static REG_CTL_3 = 0x22;
+    static REG_CTL_4 = 0x23;
+    static REG_STATUS = 0x27;
+    static REG_INT_CFG = 0x30;
+    static REG_INT_SRC = 0x31;
+    static REG_INT_THS_L = 0x32;
+    static REG_INT_THS_H = 0x33;
+    static SENSITIVITY_OF_MIN_SCALE = 27368.0; // = (4 gauss scale) * (6842 LSB/gauss at 4 gauss scale)
+
+    _i2c = null;
+    _address = null;
+    _scale = null;
+
+    function constructor(i2c, address=0x1C) {
+        _i2c = i2c;
+        _address = address;
+
+        init();
+    }
+
+    function init() {
+        // Update the cached scale so that we can convert readings to gauss
+        local reg2 = _readRegister(REG_CTL_2);
+        _scale = math.pow(4 , (reg2 >> 5) + 1);
+
+        // This is not a device default, but is convenient
+        setConversionType(CONVERSION_TYPE_CONTINUOUS);
+    }
+
+    function enable(state) {
+        local bits = state ? 0x00 : 0x02;
+        _writeRegister(REG_CTL_3, bits, 0x02);
+    }
+
+    function setPerformance(performanceRating) {
+        local bitsXY = performanceRating << 5;
+        _writeRegister(REG_CTL_1, bitsXY, 0x60);
+
+        local bitsZ = performanceRating << 2;
+        _writeRegister(REG_CTL_1, bitsZ, 0x0C);
+    }
+
+    function setDataRate(dataRate) {
+        local bits = 0x00;
+        if (dataRate == DATA_RATE_FAST) {
+            bits = 0x02;
+            return DATA_RATE_FAST;
+        } else {
+            // Cap the data rate before feeding it to equation
+            if (dataRate > 80) {
+                dataRate = 80;
+            }
+            // This is the equation used to convert data rates to the proper bitfield
+            bits = (math.log(dataRate / 0.625) / math.log(2)).tointeger() << 2;
+        }
+        _writeRegister(REG_CTL_1, bits, 0x1E);
+
+        // Return actual rate used
+        return 0.625 * math.pow(2, (bits >> 2));
+    }
+
+    function setConversionType(conversionType) {
+        _writeRegister(REG_CTL_3, conversionType, 0x01);
+    }
+
+    function setScale(scale) {
+        _scale = scale;
+
+        // Cap the scale before sending it to equation
+        if (_scale < 4) {
+            _scale = 4;
+        }
+
+        if (_scale > 16) {
+            _scale = 16;
+        }
+
+        local bits = ((_scale / 4) - 1) << 5;
```

Comment:
may need to call tointeger() here to prevent this from throwing a runtime error if the user passes in a floating-point scale.